stan/math/prim/mat/fun/ocl_gpu.hpp

```diff
@@ -0,0 +1,296 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_OCL_GPU_HPP
+#define STAN_MATH_PRIM_MAT_FUN_OCL_GPU_HPP
+
+#define __CL_ENABLE_EXCEPTIONS
+
+#include <stan/math/prim/arr/err/check_opencl.hpp>
+#include <stan/math/prim/scal/err/domain_error.hpp>
+#include <CL/cl.hpp>
+#include <cmath>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#define DEVICE_FILTER CL_DEVICE_TYPE_GPU
+
+/*
+ *    @file stan/math/prim/mat/fun/ocl.hpp
+ *    @brief Initialization for OpenCL: find platforms, devices,
+ *      create context, command queue etc.
+ */
+
+namespace stan {
+namespace math {
+
+inline void compile_kernel_group(std::string group);
+
+static int initialized_ = 0;
+typedef std::map<std::string, std::string> map_string;
+typedef std::map<std::string, cl::Kernel> map_kernel;
+typedef std::map<std::string, bool> map_bool;
+static map_string kernel_groups;
+static map_string kernel_strings;
+static map_kernel kernels;
+static map_bool compiled_kernels;
+static std::string dummy_kernel =
+  "__kernel void dummy(__global const int* foo) { };";
+
+/**
+ * Initalizes the global std::map variables that
+ * hold the OpenCL kernel sources, the groups to
+ * which each kernel is assigned to and the
+ * information about which kernel was already compiled.
```

Comment:
Add doc. This really needs to indicate what side-effects it has. Or design it so it can be tested.